### DISCUSSION TITLE
Binaries for the top-level package

### DISCUSSION THREAD

**Discussion by mmomtchev** [xpack/xpm-js#197] [Q&A]
Any ideas how can I (ab)use the binaries distribution system to have precompiled binaries?

I have a package that defines both `xpack.actions` and `xpack.binaries`. I want to be able to make an npm install script that launches a `xpm` command to download the binaries for the platform and if that fails, launches the build action. Currently, by looking at the code, I have the impression that binaries are downloaded only for the dependencies, never for the top package?

**Comment by ilg-ul** [comment]
> I want to be able to make an npm install script that launches a xpm command to download the binaries for the platform and if that fails, launches the build action.

That would be an interesting achievement!

> I have the impression that binaries are downloaded only for the dependencies, never for the top package

In my use case, the binaries are used as build tools, so they belong to devDependencies. 

If you have ideas how to improve things, sure, you can suggest new features.

**Reply by mmomtchev** [reply]
Still hesitating, because I just realized that if there are precompiled binaries, the best course of actions is to not install the xpacks at all

**Reply by ilg-ul** [reply]
> if there are precompiled binaries, the best course of actions is to not install the xpacks at all

I don't know what you mean, source xPacks are similar to npm modules, just that you can use any language you want, they are not specific to JS/TS.

One such example is [utils-lists-xpack](https://github.com/micro-os-plus/utils-lists-xpack/), which you can add as a dependency to a C++ project.

Precompiled binaries are used to run the builds and especially to build and run the tests. They are indeed plain binary archives, which can be extracted directly, without any special tools, but xpm automates this process, in a platform independent way, making things much more convenient.

The selection of such tools is limited to [xpack-dev-tools](https://github.com/xpack-dev-tools/), but these tools are enough for my tests, and, that's interesting, to building the tools themselves, since each tool build script uses previous versions of these tools.

**Reply by mmomtchev** [reply]
I mean, `xpm install` is a bad place to install the binaries - because it will also pull all the xpacks - which won't be needed if I am going to install a precompiled binary. I would like to fallback on `xpm` only if the project is being rebuilt from source. A new feature in `xpm`, something like `xpm install-binaries` would solve this - but I think I will just reuse [`prebuild-install`](https://www.npmjs.com/package/prebuild-install) - which already exists.

**Reply by ilg-ul** [reply]
I'm afraid I cannot follow you... :-(
